sync_module/selective_sync/selective_sync.go

```diff
@@ -0,0 +1,167 @@
+	}
+	if cfg.VisionService == "" {
+		return nil, fmt.Errorf(`expected "vision_service" attribute in %q`, path)
+	}
+
+	return []string{cfg.DataManager, cfg.Camera, cfg.VisionService}, nil
+}
+
+type visionSyncer struct {
+	resource.Named
+	actualDM      datamanager.Service
+	camera        camera.Camera
+	visionService vision.Service
+
+	cancelCtx  context.Context
+	cancelFunc func()
+	logger     golog.Logger
+}
+
+func newSelectiveSyncer(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger golog.Logger) (resource.Resource, error) {
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	v := &visionSyncer{
+		Named:      conf.ResourceName().AsNamed(),
+		logger:     logger,
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+	}
+	if err := v.Reconfigure(ctx, deps, conf); err != nil {
+		return nil, err
+	}
+	v.startBackgroundProcess()
+	return v, nil
+}
+
+func (s *visionSyncer) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
+	dmConfig, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+
+	s.actualDM, err = datamanager.FromDependencies(deps, dmConfig.DataManager)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get dm %v for visionSyncer", dmConfig.DataManager)
+	}
+	// If sync_disabled is false for the data manager, there are no guarantees that the selective sync
+	// behavior will function as intended. It is possible that data will be synced irregardless of the trigger.
+	s.logger.Warnw("Behavior is undetermined when scheduled sync is enabled and using selective sync", "data manager", dmConfig.DataManager)
+	s.camera, err = camera.FromDependencies(deps, dmConfig.Camera)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get camera %v for visionSyncer", dmConfig.VisionService)
+	}
+	s.visionService, err = vision.FromDependencies(deps, dmConfig.VisionService)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get vision service %v for visionSyncer", dmConfig.VisionService)
+	}
+
+	return nil
+}
+
+// DoCommand simply echos whatever was sent.
+func (s *visionSyncer) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
+	return cmd, nil
+}
+
+func (s *visionSyncer) startBackgroundProcess() {
+	utils.PanicCapturingGo(func() {
+		stream, err := s.camera.Stream(s.cancelCtx)
+		defer utils.UncheckedErrorFunc(func() error {
+			return stream.Close(s.cancelCtx)
+		})
+		if err != nil {
+			s.logger.Error("could not get camera stream")
+			return
+		}
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				// Only return captured image if it contains a certain color set by the vision service.
+				if err != nil {
+					s.logger.Error("could not get next source image")
+					return
+				}
+				// Check for stuff, if true Sync
+				img, rel, err := stream.Next(s.cancelCtx)
+				defer rel()
```

Comment:
I'm not too familiar with `stream.Next` - is it necessary to defer the `rel()` in a for loop here? this will cause those deferred calls to stack up until the program ends, and only then be called. This is generally not great because it leads to constantly increasing memory usage here which would eventually cause issues (even if it's not inherently problematic for `stream` for `rel()` to not be called for a long time)

Reply:
Oh no, I very much did not mean to defer `release()`, this can be called right after we're done using the image.